[CoordinatedGraphics] Compute the layers tile size instead of using a fixed value

carlosgcampos opened the pull request:
#### 51a8c38ce20633b24ff72b87c2e3429e7cfc69f5
<pre>
[CoordinatedGraphics] Compute the layers tile size instead of using a fixed value
<a href="https://bugs.webkit.org/show_bug.cgi?id=300093">https://bugs.webkit.org/show_bug.cgi?id=300093</a>

Reviewed by Miguel Gomez.

This way we can use a different logic depending on whether we are using
CPU or GPU rendering for the tiles. In the case of CPU rendering we use
smaller tiles of 256x256 by default, while for GPU rendering we use the
visible size for the width and 1/4 of the visible height, similar to what
chromium does.

Canonical link: <a href="https://commits.webkit.org/301398@main">https://commits.webkit.org/301398@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e19011dfc78ca3c3ee3abf97ae178d6ed033f03

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125694 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45356 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36105 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132559 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77578 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127565 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46040 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53914 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95759 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63878 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128642 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36806 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112399 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76251 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35708 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30583 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76027 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106584 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30798 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135234 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52484 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40244 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104224 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52931 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108611 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103952 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49312 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27625 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49723 "Hash 4e19011d for PR 51742 does not build (failure)") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19693 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52379 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58185 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51727 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55078 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53423 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreProxy.cpp:
```diff
@@ -363,6 +330,78 @@ void CoordinatedBackingStoreProxy::createOrDestroyTiles(const IntRect& visibleRe
+        if (!overridenTileSize) {
+            auto tokens = String::fromUTF8(tileSizeEnv).split('x');
+            if (!tokens.isEmpty()) {
+                int width = parseInteger<int>(tokens[0]).value_or(0);
```
s/int/unsigned/

carlosgcampos replied:
These are used to create a IntSize that expects int

nikolaszimmermann replied:
But then you need to test that it's positive? I'd rather parseInteger<unsigned> and implicitly cast to int when constructing the IntSize -- then parseInteger does the positivity check for you.

carlosgcampos replied:
We already check it with the not empty check. I don't think we should complicate this for a debugging only env var.

nikolaszimmermann replied:
Well I didn't know that isEmpty checks that. Fine with me.

nikolaszimmermann replied:
I did just remember that I used `parseInteger<unsigned>` in `SkiaPaintingEngine` when I wanted positive values. But I didn't have to feed an Intsize with it.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreProxy.cpp:
```diff
@@ -363,6 +330,78 @@ void CoordinatedBackingStoreProxy::createOrDestroyTiles(const IntRect& visibleRe
+    if (tileSizeEnv && *tileSizeEnv) {
+        static std::optional<IntSize> overridenTileSize;
+        if (!overridenTileSize) {
+            auto tokens = String::fromUTF8(tileSizeEnv).split('x');
```
I would use `StringView::fromLatin1(tileSizeEnv)`, not the UTF-8 variant.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreProxy.cpp:
```diff
@@ -363,6 +330,78 @@ void CoordinatedBackingStoreProxy::createOrDestroyTiles(const IntRect& visibleRe
+// This is based on the same heuristics used by chromium to compute the tile size.
+IntSize CoordinatedBackingStoreProxy::computeTileSize(int32_t maxTextureSize) const
+{
+    static const char* tileSizeEnv = getenv("WEBKIT_LAYERS_TILE_SIZE");
```
I would suggest not to make those static, but wrap this in a `std::call_once` block.

```
    static std::once_flag onceFlag;
    std::call_once(onceFlag, [] {
...
```

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreProxy.h:
```diff
@@ -126,13 +125,12 @@ class CoordinatedBackingStoreProxy final : public ThreadSafeRefCounted<Coordinat
-    void reset();
     void invalidateRegion(const Vector<IntRect, 1>&);
-    void createOrDestroyTiles(const IntRect& visibleRect, const IntRect& scaledContentsRect, float coverAreaMultiplier, Vector<uint32_t>& tilesToCreate, Vector<uint32_t>& tilesToRemove);
+    void createOrDestroyTiles(const IntRect& unscaledVisibleRect, const IntRect& unscaledContentsRect, float contentsScale, int32_t maxTextureSize, Vector<uint32_t>& tilesToCreate, Vector<uint32_t>& tilesToRemove);
```
Why int32_t / uint32_t here? int/unsigned is what we'd normally use, no?

carlosgcampos replied:
int32_t is because that's what glGetIntegerv returns, the uint32_t was there before.

nikolaszimmermann replied:
But what is our policy on that? Do we want that kind of mix?

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
```diff
@@ -471,6 +471,7 @@ void CoordinatedPlatformLayer::setContentsScale(float contentsScale)
         return;
 
     m_contentsScale = contentsScale;
+    m_needsTilesUpdate = true;
```
Oh, did we miss that before?

nikolaszimmermann replied:
Ah ic, you changed the logic below in updateBackingStore.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.h:
```diff
@@ -181,6 +182,7 @@ class CoordinatedPlatformLayer : public ThreadSafeRefCounted<CoordinatedPlatform
     bool isCompositionRequiredOrOngoing() const;
     void requestComposition();
     RunLoop* compositingRunLoop() const;
+    int32_t maxTextureSize() const;
```
I'd go for `int` everywhere.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
```diff
@@ -131,6 +132,7 @@ class ThreadedCompositor : public ThreadSafeRefCounted<ThreadedCompositor>, publ
     std::unique_ptr<WebCore::GLContext> m_context;
 
     bool m_flipY { false };
+    int32_t m_maxTextureSize { 0 };
```
Is it guaranteed that this is always initialized to a non-zero value before we query it?
Do we need to assert that we don't report a 0 texture size or, use a sensible default? e.g. 1024?

carlosgcampos replied:
Good point, I think we should handle that instead, because we can also end up with a 0 max size if the layer is queried when it doesn't have a client (which should never happen).

nikolaszimmermann reviewed (commented):
Looks good in general, some style comments.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreProxy.cpp:
```diff
@@ -363,6 +334,84 @@ void CoordinatedBackingStoreProxy::createOrDestroyTiles(const IntRect& visibleRe
+            width += 2;
+            height += 2;
+
+            // Round GPU default tile sizes to a multiple of 32 to prevent rounding errors.
```
Let's see if that solves is ;-) Do you have more insights in the "rounding errors" now?

carlosgcampos replied:
I have no idea, tbh, I added it just to try, because I can't reproduce the failures locally.

carlosgcampos replied:
It didn't help

carlosgcampos committed:
[CoordinatedGraphics] Compute the layers tile size instead of using a fixed value
https://bugs.webkit.org/show_bug.cgi?id=300093

Reviewed by Miguel Gomez.

This way we can use a different logic depending on whether we are using
CPU or GPU rendering for the tiles. In the case of CPU rendering we use
smaller tiles of 256x256 by default, while for GPU rendering we use the
visible size for the width and 1/4 of the visible height, similar to what
chromium does.

Canonical link: https://commits.webkit.org/301398@main
